_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,223 @@
+
+Having the familiarity with the overall layout and the goal of our Fraud Detection engine, let's now go into the details of what is required to implement such system.
+
+### Dynamic Data Partitioning
+
+The first pattern we will look into is Dynamic Data Partitioning.
+
+Flink's DataStream API includes an important operator that determines the distribution of incoming data (shuffle) among the parallel computational nodes of the cluster - the **keyBy()** operator. It partitions the stream in such a way that all records with the same key are assigned to the same partition and are hence sent to the same physical instance of the subsequent "keyed" operator.
+
+In a typical statically-defined streaming application the choice of the key is fixed and it is usually immediately derived from one of the fields of the incoming records. For instance, in case of a simple window-based computation over a stream of Tuples, we could use the first field of type String (denoted by 0) as the key:
+
+```java
+DataStream<Tuple2<String, Integer>> input = // [...]
+DataStream<...> windowed = input
+  .keyBy(0)
+  .window(/*window specification*/);
+```
+
+This approach is the basic building block for achieving horizontal scalability in a wide range of use cases. However, in the case of an application that strives to provide flexibility in business logic at runtime, this is not enough.
+To understand why this is the case, let us start with articulating a realistic sample rule definition for our fraud detection system in a form of a functional requirement:  
+
+*"Whenever the **sum** of accumulated **payment amount** from the same **beneficiary** to the same **payee** within the **duration of a week** is **greater** than **1 000 000 $** - fire an alert."*
+
+In this formulation we can spot a number of parameters that we would like to be able to specify in a newly submitted rule and possibly even later modify or tweak at runtime:
+
+1. Aggregation field (payment amount)  
+1. Grouping fields (beneficiary + payee)  
+1. Aggregation function (sum)  
+1. Window duration (1 week)  
+1. Limit (1 000 000)  
+1. Limit operator (greater)  
+
+Accordingly, we will use the following simple JSON format to define the aforementioned parameters:
+
+```json  
+{
+  "ruleId": 1,
+  "ruleState": "ACTIVE",
+  "groupingKeyNames": ["beneficiaryId", "payeeId"],
+  "aggregateFieldName": "paymentAmount",
+  "aggregatorFunctionType": "SUM",
+  "limitOperatorType": "GREATER",
+  "limit": 1000000,
+  "windowMinutes": 10080
+}
+```
+
+At this point it is important to understand that **`groupingKeyNames`** determine the actual physical grouping of events - all Transactions with the same values of specified parameters (e.g. _beneficiary #25 -> payee #12_ have to be aggregated in the same physical instance of the evaluating operator. Naturally, the process of distributing data in such a manner in Flink's API is realised by a `keyBy()` function.
```

Comment:
```suggestion
At this point, it is important to understand that **`groupingKeyNames`** determine the actual physical grouping of events — all Transactions with the same values of specified parameters (e.g. _beneficiary #25 -> payee #12) have to be aggregated in the same physical instance of the evaluating operator. Naturally, the process of distributing data in such a way in Flink's API is realised by a `keyBy()` function.
```

Reply:
Second underscore is for italic formatting.

Reply:
Okay, and where should the parenthesis close? Is it after the italic formatting?